cpp/src/arrow/acero/hash_join_node_test.cc

```diff
@@ -3253,5 +3252,221 @@ TEST(HashJoin, ManyJoins) {
+    Declaration join{"hashjoin", {std::move(left), std::move(right)}, join_options};
+
+    ASSERT_OK_AND_ASSIGN(auto actual, DeclarationToExecBatches(std::move(join)));
+
+    AssertExecBatchesEqual(expected.schema, expected.batches, actual.batches);
+    AssertSchemaEqual(expected.schema, actual.schema);
+  }
+
+  // Testing large_list(int32())
+  {
+    auto l_schema = schema({field("l_i32", int32()),
+      field("l_list", large_list(int32()))});
+    auto r_schema = schema({field("r_i32", int32()),
+      field("r_list", large_list(int32()))});
+
+    std::vector<FieldRef> l_keys{{"l_i32"}};
+    std::vector<FieldRef> r_keys{{"r_i32"}};
+
+    BatchesWithSchema l_batches = GenerateBatchesFromString(l_schema, {
+      R"([
+        [1, [4, 5, 6, 7, 8]],
+        [2, [100, 200, 300]],
+        [3, null],
+        [4, [123789, 43289]],
+        [null, []]
+      ])"
+    });
+    BatchesWithSchema r_batches = GenerateBatchesFromString(r_schema, {
+      R"([
+        [1, [-1, -2]],
+        [2, [400]],
+        [3, [null, 6969]],
+        [null, null]
+      ])"
+    });
+
+    BatchesWithSchema expected;
+    expected.batches = {
+        ExecBatchFromJSON({int32(), large_list(int32()), int32(), large_list(int32())}, R"([
+      [1, [4, 5, 6, 7, 8], 1, [-1, -2]],
+      [2, [100, 200, 300], 2, [400]],
+      [3, null, 3, [null, 6969]]
+    ])")};
+
+    expected.schema = schema({field("l_i32", int32()),
+      field("l_list", large_list(int32())), field("r_i32", int32()),
+      field("r_list", large_list(int32()))});
+
+    AsyncGenerator<std::optional<ExecBatch>> sink_gen;
+
+    HashJoinNodeOptions join_options{JoinType::INNER, l_keys, r_keys};
+    Declaration left{"source",
+                      SourceNodeOptions{l_batches.schema, l_batches.gen(parallel, slow)}};
+    Declaration right{"source",
+                      SourceNodeOptions{r_batches.schema, r_batches.gen(parallel, slow)}};
+    Declaration join{"hashjoin", {std::move(left), std::move(right)}, join_options};
+
+    ASSERT_OK_AND_ASSIGN(auto actual, DeclarationToExecBatches(std::move(join)));
+
+    AssertExecBatchesEqual(expected.schema, expected.batches, actual.batches);
+    AssertSchemaEqual(expected.schema, actual.schema);
+  }
+
+  // Testing if we have a scalar list(int32()) column
+  {
+    auto l_schema = schema({field("l_i32", int32()), field("l_list", list(int32()))});
+    auto r_schema = schema({field("r_i32", int32()), field("r_list", list(int32()))});
+
+    BatchesWithSchema l_batches, r_batches;
+
+    auto l_batch = ExecBatch::Make({ArrayFromJSON(int32(), R"([1, 2])"),
+                    ScalarFromJSON(list(int32()), "[1, 2, null, 4]")});
+    auto r_batch = ExecBatch::Make({ArrayFromJSON(int32(), R"([1, 2, 3])"),
+                    ArrayFromJSON(list(int32()), R"([[], [1], [69]])")});
+
+    l_batches.schema = l_schema;
+    l_batches.batches.push_back(l_batch.ValueUnsafe());
```

Comment:
I think you should call `ValueOrDie`. `ValueUnsafe` will result in undefined behavior if the call failed for some reason.